File: VideoLocker/src/main/java/org/edx/mobile/module/notification/NotificationDelegate.java
package org.edx.mobile.module.notification;

import org.edx.mobile.model.api.EnrolledCoursesResponse;
import java.util.List;

/**
 *  abstracted for different implementation.
 *  it may not make much sense for now. because different notification
 *  middleware may invoke different workflow.
 *
 */
public interface NotificationDelegate {
    public void unsubscribeAll( );

    /**
     * if user login,  he should subscribe channels based on local pref setting.
     */
    public void resubscribeAll();

    //check if local subscribed is not in notification server
    //then try to subscribe it
    public void syncWithServerForFailure();

    /**
     * sync with current course enrollment.
     * based on current course enrollment, we subscribe/unsubscribe to the notification
     *  and update the local preference.
     * @param responses
     */
    public void checkCourseEnrollment(List<EnrolledCoursesResponse> responses);

    /**
     *
     * @param courseId  also the channel id
     * @param subscribe subscribe or unsubscribe to courseId channel
     */
    public void changeNotificationSetting(String courseId, String channelId, boolean subscribe);


    /**
     *
     * @param channel
     * @return
     */
    public boolean isSubscribedByCourseId(String channel);

    /**
     * app upgrade or new installation, it may need to resync with notification server
     */
    public void checkAppUpgrade();
}

File: OpenEdXMobile/src/main/java/org/edx/mobile/module/notification/NotificationDelegate.kt
package org.edx.mobile.module.notification

import org.edx.mobile.model.api.EnrolledCoursesResponse

/**
 * Abstracted for different implementation.
 * It may not make much sense for now because different notification middleware may invoke
 * different workflow.
 */
interface NotificationDelegate {
    fun unsubscribeAll()

    /**
     * If learner logins, he should subscribe channels based on local pref setting
     */
    fun resubscribeAll()

    /**
     * Checks if local subscribed is not in the notification server then try to subscribe it
     */
    fun syncWithServerForFailure()

    /**
     * Sync with current course enrollment.
     * Based on current course enrollment, we subscribe/unsubscribe to the notification
     * and update the local preference.
     *
     * @param responses enrolled courses and remote config data
     */
    fun checkCourseEnrollment(responses: List<EnrolledCoursesResponse>)

    /**
     * @param courseId the course id of the underline course
     * @param channelId the notification channel id of this course
     * @param subscribe subscribe or unsubscribe to courseId channel
     */
    fun changeNotificationSetting(courseId: String, channelId: String, subscribe: Boolean)

    /**
     *
     * @param channel the notification channel
     * @return
     */
    fun isSubscribedByCourseId(channel: String): Boolean

    /**
     * App upgrade or new installation, it may need to re-sync with notification server
     */
    fun checkAppUpgrade()
}
